dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/IntrinsicsInlining.java

```diff
@@ -0,0 +1,117 @@
+package com.dat3m.dartagnan.program.processing;
+
+import com.dat3m.dartagnan.exception.ParsingException;
+import com.dat3m.dartagnan.expression.BNonDet;
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.ExpressionFactory;
+import com.dat3m.dartagnan.expression.INonDet;
+import com.dat3m.dartagnan.expression.type.BooleanType;
+import com.dat3m.dartagnan.expression.type.IntegerType;
+import com.dat3m.dartagnan.expression.type.TypeFactory;
+import com.dat3m.dartagnan.program.Program;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.Thread;
+import com.dat3m.dartagnan.program.event.EventFactory;
+import com.dat3m.dartagnan.program.event.core.Event;
+import com.dat3m.dartagnan.program.event.functions.DirectFunctionCall;
+import com.dat3m.dartagnan.program.event.functions.DirectValueFunctionCall;
+import com.google.common.primitives.UnsignedInteger;
+import com.google.common.primitives.UnsignedLong;
+import org.sosy_lab.common.configuration.Configuration;
+import org.sosy_lab.common.configuration.InvalidConfigurationException;
+
+import java.math.BigInteger;
+import java.util.List;
+
+public class IntrinsicsInlining implements ProgramProcessor {
+
+    private int constantId;
+
+    private IntrinsicsInlining() {
+    }
+
+    public static IntrinsicsInlining fromConfig(Configuration ignored) throws InvalidConfigurationException {
+        return new IntrinsicsInlining();
+    }
+
+    @Override
+    public void run(Program program) {
+        constantId = 0;
+        for (Thread thread : program.getThreads()) {
+            run(thread);
+        }
+    }
+
+    private void run(Thread thread) {
+        for (Event event : List.copyOf(thread.getEvents())) {
+            if (!(event instanceof DirectFunctionCall call)) {
+                continue;
+            }
+            assert !call.getCallTarget().hasBody();
+
+            List<Event> replacement = switch (call.getCallTarget().getName()) {
+                case "__VERIFIER_nondet_bool",
+                        "__VERIFIER_nondet_int", "__VERIFIER_nondet_uint", "__VERIFIER_nondet_unsigned_int",
+                        "__VERIFIER_nondet_short", "__VERIFIER_nondet_ushort", "__VERIFIER_nondet_unsigned_short",
+                        "__VERIFIER_nondet_long", "__VERIFIER_nondet_ulong",
+                        "__VERIFIER_nondet_char", "__VERIFIER_nondet_uchar" -> inlineNonDet(call);
+                default -> throw new UnsupportedOperationException(
```

Comment:
I don't get this. Isn't this matching every function call (which we need!) that is not a `__VERIFIER_nondet` one?

Reply:
It is matching every other call. This pass runs after regular inlining and thread creation, so that all calls except for intrinsic one's are resolved. The pass then tries to resolve all the intrinsics and if there is one that it cannot resolve, it will throw an exception, which is fine.
While I don't think that this single pass will be able to resolve all intrinsics in the future, i.e., we will likely get rid of the exception throwing, at least for now it is correct.